pkg/filesystem/file_watcher.go

```diff
@@ -0,0 +1,126 @@
+package filesystem
+
+import (
+	"context"
+	"errors"
+	"github.com/fsnotify/fsnotify"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+const workItemKey = "key"
+
+// FileWatcher watches a single file and trigger the given handler function
+type FileWatcher struct {
+	path    string
+	handler FileContentHandler
+
+	watcher *fsnotify.Watcher
+
+	// Instead of doing the work in processEvent, a queue is used primarily to
+	// make testing easier and to keep the FileWatcher simple.  A single item
+	// will be added to the queue to denote the file should be reloaded.
+	// Additional events will be deduped until the item is removed with Done().
+	// If there is an error reloading the file, we enqueue rate limited (with a
+	// max wait of 10 seconds).  The workqueue was chosen because it allows us
+	// to deduplicate reloads and retry with rate limit on failure.  This
+	// pattern is borrowed from
+	// https://github.com/kubernetes/kubernetes/blob/3d67e162a03d0d724dc5a15a0617c5e8572c7b4a/staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamic_serving_content.go
+	queue workqueue.RateLimitingInterface
+}
+
+type FileContentHandler func(content []byte) error
+
+// NewFileWatcher creates a FileWatcher
+func NewFileWatcher(purpose string, path string, handler FileContentHandler) *FileWatcher {
+	return &FileWatcher{
+		path:    path,
+		handler: handler,
+		queue:   workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), purpose),
+	}
+}
+
+// Watch sets up the fsnotify watcher and add the file that we are interested in.  The file watcher
+// and worker run in goroutines.  The goroutines are stopped when the ctx is cancelled.
+func (f *FileWatcher) Watch(ctx context.Context) error {
+	// Trigger initial file load
+	f.queue.Add(workItemKey)
+
+	var err error
+	f.watcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+
+	go wait.UntilWithContext(ctx, f.runWorker, time.Second)
+
+	// Start listening for events.
+	go func() {
+		for {
+			select {
+			case err := <-f.watcher.Errors:
+				klog.ErrorS(err, "Error from watcher")
+			case e := <-f.watcher.Events:
+				klog.V(3).InfoS("Event received", "event", e)
+				f.processEvent(e)
+			case <-ctx.Done():
+				klog.Info("context closed, stopping FileWatcher")
+				f.watcher.Close()
+				return
+			}
+		}
+	}()
+
+	dir := filepath.Dir(f.path)
+	err = f.watcher.Add(dir)
+	if err != nil {
+		klog.Fatal(err)
+	}
+
+	return nil
+}
+
+// processEvent adds an item to the workqueue.
+func (f *FileWatcher) processEvent(event fsnotify.Event) {
+	if event.Name == f.path {
+		f.queue.Add(workItemKey)
+	}
+}
+
+func (f *FileWatcher) runWorker(ctx context.Context) {
+	for f.processNextWorkItem(ctx) {
+	}
+}
+
+func (f *FileWatcher) processNextWorkItem(ctx context.Context) bool {
```

Comment:
Can we make this more explicit? Some nice enum that says continue, stop polling?

I also wonder. What if we have an event that is always failing, when will we stop processing it? We always re-queue it. We will spend all of our "rate" buckets on a useless message until it goes away. Is that desirable?

What if we have really slow I/O s.t. it makes this function run slower than the refill rate, wouldn't that make this go into an infinite loop, just wasting resources? I am thinking of a corrupted disk. In that scenario, I/O is really slow and part of the file might be allocated to two disk sectors. If the second one is corrupted, the OS will error out and we will keep retrying.

Reply:
Updating the naming.

For the re-queueing, i think the current implementation is okay.  The code is using a rate limited queue, which uses exponential backoff upon error.